### PR TITLE
Truncated Signed Distance Value Converter

### PR DESCRIPTION
**Pull request by kdaun**
Introduces conversions between float and uint16 representations for truncated signed distance values and weights.

### CODE REVIEW

File: cartographer/mapping/tsd_value_converter.h [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    DCHECK_GE(value, 1);
+    DCHECK_LE(value, 32767);
+    return value;
+  }
+
+  // Converts a uint16 (which may or may not have the update marker set) to a
+  // probability in the range [kMinTSD, kMaxTSD].
+  inline float ValueToTSDF(const uint16 value) const {
+    return kValueToTSDF[value];
+
```
Comment (gaschler) [inline cartographer/mapping/tsd_value_converter.h]: drop empty line

File: cartographer/mapping/tsd_value_converter.h [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+namespace mapping {
+
+// Provides conversions between float and uint16 representations for
+// truncated signed distance values and weights.
+class TSDValueConverter {
+ public:
+  TSDValueConverter(float max_tsdf, float max_weight);
+
+  // Clamps probability to be in the range [kMinTSD, kMaxTSD].
+  inline float ClampTSD(const float tsd) const {
```
Comment (gaschler) [inline cartographer/mapping/tsd_value_converter.h]: if these are only used by this class, make them private

File: cartographer/mapping/tsd_value_converter.h [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    return common::Clamp(tsd, kMinTSD, kMaxTSD);
+  }
+
+  // Clamps probability to be in the range [kMinTSD, kMaxTSD].
```
Comment (gaschler) [inline cartographer/mapping/tsd_value_converter.h]: probability?

File: cartographer/mapping/tsd_value_converter.h [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+  inline float ClampWeight(const float tsd) const {
+    return common::Clamp(tsd, kMinWeight, kMaxWeight);
+  }
+
+  // Converts a tsdf to a uint16 in the [1, 32767] range.
+  inline uint16 TSDFToValue(const float tsd) const {
+    const int value = common::RoundToInt((ClampTSD(tsd) - kMinTSD) *
+                                         (32766.f / (kMaxTSD - kMinTSD))) +
```
Comment (gaschler) [inline cartographer/mapping/tsd_value_converter.h]: save the divisor as a constant to re-use can potentially save 2 flops

File: cartographer/mapping/tsd_value_converter.h [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+  float getMaxWeight() const { return kMaxWeight; }
+  float getMinWeight() const { return kMinWeight; }
+
+ private:
+  float SlowValueToTSD(const uint16 value) const;
+  std::vector<float> PrecomputeValueToTSD();
+  float SlowValueToWeight(const uint16 value) const;
+  std::vector<float> PrecomputeValueToWeight();
+
+  float kMaxTSD;
```
Comment (gaschler) [inline cartographer/mapping/tsd_value_converter.h]: style: max_tsd_, also the other ones

File: cartographer/mapping/tsd_value_converter.cc [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    // Unknown cells have kMinProbability.
+    return kMinTSD;
+  }
+  const float kScale = (kMaxTSD - kMinTSD) / 32766.f;
+  return value * kScale + (kMinTSD - kScale);
+}
+
+std::vector<float> TSDValueConverter::PrecomputeValueToTSD() {
+  std::vector<float> result;
+  // Repeat two times, so that both values with and without the update marker
```
Comment (gaschler) [inline cartographer/mapping/tsd_value_converter.cc]: reserve size

File: cartographer/mapping/tsd_value_converter.cc [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+  // can be converted to a tsd.
+  for (int repeat = 0; repeat != 2; ++repeat) {
```
Comment (gaschler) [inline cartographer/mapping/tsd_value_converter.cc]: This is not "obviously correct".
Perhaps you can just iterate over all uint16, and AND away the marker?

File: cartographer/mapping/tsd_value_converter.h [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+  inline float ValueToTSDF(const uint16 value) const {
+    return kValueToTSDF[value];
+
+  }  // Converts a uint16 (which may or may not have the update marker set) to a
+  // probability in the range [kMinWeight, kMaxWeight].
+  inline float ValueToWeight(const uint16 value) const {
+    return kValueToWeight[value];
+  }
+
+  uint16 getUnknownTSDFValue() const { return kUnknownTSDValue; }
```
Comment (gaschler) [inline cartographer/mapping/tsd_value_converter.h]: these three can be static

File: cartographer/mapping/tsd_value_converter.cc [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    for (int value = 0; value != 32768; ++value) {
+      result.push_back(SlowValueToWeight(value));
+    }
+  }
+  return result;
+}
+
+}  // namespace mapping
+}  // namespace cartographer
```
Comment (gaschler) [inline cartographer/mapping/tsd_value_converter.cc]: super nit: add new line at end of files

File: cartographer/mapping/tsd_value_converter.cc (hunk 7 above)
Comment (kdaun) [reply cartographer/mapping/tsd_value_converter.cc]: Yes, great idea.

**Comment by kdaun** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user kdaun. Merge queue now has a length of 1.

**Pull request merged**